README
===========================================
RAZR SOFTWARE PACKAGE FOR MATHWORK'S MATLAB
===========================================

RAZR - Room acoustics simulator

Version 0.93

Copyright (c) 2014-2017, Torben Wendt, Steven van de Par, Stephan Ewert,
University of Oldenburg, Germany.
All rights reserved. 

Author: Torben Wendt
e-mail: torben.wendt(at)uni-oldenburg.de


===============================================================================
Algorithm reference
===============================================================================

Wendt, T., van de Par, S., Ewert, S.D. (2014): A Computationally-Efficient and
Perceptually-Plausible Algorithm for Binaural Room Impulse Response Simulation.
J. Audio Eng. Soc., Vol. 62, No. 11, pp. 748-766.

Please note: The set of default options contain some optimizations compared to
the setup described in the publication above. You can however load the original
setup by using GET_DEFAULT_OPTIONS_2014.M. To see how to use options, see also
the subsection "Getting started/Data structures", below in this file.


===============================================================================
What's in the distribution?
===============================================================================

This distribution adds a toolbox for synthesizing and analyzing binaural
room impulse responses (BRIRs) to your Matlab installation. Example simulations
are included.


Included Files
--------------

./*                     - RAZR top-level functions, README, license, disclaimer
ANALYSIS_TOOLS/
    *.M                 - Collection of tools to analyze BRIRs
    HEADPHONE_EQ/*.mat  - Equalization files for headphones
EXAMPLES/*.M            - Some examples to create BRIRs
BASE/**                 - RAZR core files
    EXTERNAL/**         - Files created by other Authors, see Credits below
    HRTF/**             - Supplementary files for HRTF application
    MEXFDN/*            - Compiled FDN mainloop
    TOOLS/*.M           - Supplementary functions for simulation and analysis

All these files must not be changed as also not permitted by RAZR's license,
unless not stated otherwise. Changes might moreover harm proper function of
this distribution. For exceptions see the following section "License and
permissions".


===============================================================================
License and permissions
===============================================================================

Unless otherwise stated, the RAZR distribution, including all files is licensed
under Creative Commons Attribution-NonCommercial-NoDerivs 4.0 International
(CC BY-NC-ND 4.0).
In short, this means that you are free to use and share (copy, distribute and
transmit) the RAZR distribution under the following conditions:

Attribution - You must attribute the RAZR distribution by acknowledgement of
              the author if it appears or if was used in any form in your work.
              The attribution must not in any way that suggests that the author
              endorse you or your use of the work.

Noncommercial - You may not use RAZR for commercial purposes.
 
No Derivative Works - You may not alter, transform, or build upon RAZR.


Exceptions are:
- The files ./BASE/EXTERNAL/**, see their respective license.

Users are explicitly granted to modify, derive their own modifications and to
distribute their modifications for the configurations, options and examples:
- The file RAZR_CFG.M in the main folder
- All files in the folder ./EXAMPLES

For further exceptions, see the license headers in the respecive files and the
section "Credits" in this file.


===============================================================================
Installation
===============================================================================

Requirements
------------
- Windows 95/98, Windows NT/2000/XP or higher, or MAC OSX, or Linux
- Mathwork's MATLAB Version 7.4 or higher


Recommended
-----------
- Sound device
- Headphones


Installation
------------
- Unzip all included files to your preferred directory.


===============================================================================
Getting Started
===============================================================================

Demo and Example scripts
------------------------
To start the demo, type "razr demo".

Besides, in the directory "EXAMPLES", you can find some example scripts, that 
demonstrate the usage of this program package. See the respective files for
details. When you have read and tried out all examples, you know the most
important features of RAZR.
To start any of the examples, type "razr addpath" to temporally add all required
folders to MATLAB's path. Then type, e.g., "example_default" or
"example_options". 

A formal usage description is provided in the documentation of the top-level
function razr(). Type "razr help" to display it.

Please see also the changelog if you have already used earlier versions of RAZR.


Data structures
---------------
RAZR uses three essential data structures, that occur in the probably most
common usage of RAZR:

ir = razr(room, op);

(1) A room is defined as a Matlab structure, called "room" in most scripts and
functions. It contains information about the room size, wall absorption,
positions of source and receiver, and other specifications. For the required
fields of a room structure see the example rooms defined in
EXAMPLES/GET_ROOM_*.M. Both, required and optional fields are explained in the
help entry of RAZR.M. Depending on your taste and needs, you can
generate and store own rooms either in functions like the examples, as *.mat
files, directly within scripts, or by any other means. For example, for
automated RIR creation for many rooms in a loop, .mat files might be the most
convenient way to store rooms as they can easily be loaded by filename.

(2) Options for the BRIR synthesis are defined in a structure, called "op" in
most scripts and functions. All possible fields and their default values are
stored in BASE/GET_DEFAULT_OPTIONS.M. To display options, check out also the
function "sop".
Some of them are experimental, refer to features in development (and thus not
contained in the current version), or might read unclear. In doubt, stick to
the default or contact the author. See EXAMPLES/EXAMPLE_OPTIONS.M to see how
to use options.

(3) Room impulse responses are stored as a structure, called "ir" in most
scripts and functions. It contains the time signal itself and some metadata,
depending on your chosen options. For instance, ir can contain also the signals
for direct sound, early and late reflections in separate fields, if the option
"return_rir_parts" is set to "true".
Most of the functions in ANALYSIS_TOOLS expect a BRIR in the format of such a
structure as input parameter.

For further information, type "razr help".


Configuration
-------------
In contrast to the options structure, which specifies how a BRIR is created, in
the configuration it is mainly defined, where RAZR finds certain files on your
computer. Hence, on different computers, the same options can be used, whereas
the configuration has to be adapted.
Configuration structures are defined in RAZR_CFG_DEFAULT.M and RAZR_CFG_USER.M.
(The latter one is automatically created when you run razr for the first time.)
Internally, RAZR_CFG_USER overloads RAZR_CFG_DEFAULT. You can also create and
store several configuration files in parallel. In the file SELECT_RAZR_CFG.M,
you can specify what configuration is actually used to override
RAZR_CFG_DEFAULT. RAZR_CFG_USER is the predefined one.


===============================================================================
Using HRTFs/HRIRs
===============================================================================

This program package does not contain any HRTF data. In the default settings, a
spherical head model (SHM) is used, implemented after [2]. However, though the
SHM yields reasonably good results (and is computationally very efficient),
using HRTFs is an essential part of BRIR simulation. If you would like to use
HRTFs you need to get a database from external sources. To link an HRTF database
to RAZR, download the database to the preferred directory on your harddrive.
Then, this database must be specified in ./RAZR_CFG.M as described in the
following:

SOFA format
-----------
RAZR supports HRIRs that are stored in the SOFA format [3], using the SOFA con-
vention "SimpleFreeFieldHRIR" [3]. If you would like to use SOFA-formatted HRIRs
you need to download the SOFA API for Matlab/Octave from

    https://www.sofaconventions.org/mediawiki/index.php/Software_and_APIs

Store and unzip the API in your preferred directory. In RAZR, open the file
RAZR_CFG_USER and set the field "cfg.sofa_api_path" to the path where you stored
the API.
Then, there are two ways to use a SOFA database in RAZR:
(1) Specify the name of the .sofa file directly as option, e.g.,
    op.hrtf_database = path/to/database.sofa;
(2) Use a shortcut. In ./RAZR_CFG.M, a shortcut to a .sofa file can be linked
    as, e.g,.:
    cfg.sofa_file__<shortcut> = path/to/database.sofa;
    (Note the double underscore before <shortcut>.) Then, the option to use this
    Database can be set as:
    op.hrtf_database = 'shortcut.sofa';
    (The suffix ".sofa" is required although it is not part of the shortcut.)

Other formats
-------------
For an HRTF database that is stored in any other format, its path on your hard-
drive must be specified in ./RAZR_CFG.M:

cfg.hrtf_path__<shortcut> = path/to/database;

where <shortcut> is the unique key for the database. For already supported data-
bases, configuration fields have already been prepared there. (Note: Some of
them are not open access databases or not published, yet.)

If you would like to use an unsupported (non-SOFA) HRTF database, you need
create two Matlab functions, which will be executed internally in RAZR:
  - hrtf_params_<shortcut>  to set database-specific parameters.
  - pick_hrir_<shortcut>    to pick a set of HRIRs from the database according
                            to a set of desired azimuth and elevation angles.
For the syntax of these functions and required parameters please see the
examples provided in ./BASE/HRTF.

If you create these functions for an open-access database, be cordially invited
to share them with the author, such that they can be included in a future
release.

All formats (SOFA and non-SOFA)
-------------------------------
To actually use an HRTF database (SOFA or non-SOFA), the following option must
be set and passed to razr:
op.spat_mode = 'hrtf';

Note that the sampling rate of the HRIRs from a database must match the sampling
rate of the BRIR generated by RAZR (44100 Hz). If necessary, it can be adjusted
using the option op.fs, e.g.,
op.fs = 48000;

HRTF usage is also demonstrated in EXAMPLES/EXAMPLE_HRTF.M which is precon-
figured for the FABIAN database [4]. To run this example with the FABIAN data-
base, download the database from [3] and store it in a preferred directory on
your harddrive. Then replace 'path_to_database' in ./RAZR_CFG.M in the SOFA file
definition for the FABIAN database with the path to the directory where you
stored the database.


===============================================================================
References
===============================================================================

[1] Wendt, T., van de Par, S., Ewert, S.D. (2014): A Computationally-Efficient
    and Perceptually-Plausible Algorithm for Binaural Room Impulse Response
    Simulation. J. Audio Eng. Soc., Vol. 62, No. 11
[2] C. P. Brown and R.O. Duda (1998): "A Structural Model for Binaural  Sound
    Synthesis," IEEE Trans. Speech Audio Processing, vol. 6, no. 5
[3] https://www.sofaconventions.org
[4] Brinkmann, F., Lindau, A., Weinzierl, S., Geissler, G., & van de Par, S.
    (2013). A high resolution head-related transfer function database including
    different orientations of head above the torso. AIA-DAGA 2013.

===============================================================================
Credits
===============================================================================

This work was supported by
- DFG-FOR 1732 "Individualisierte Hoerakustik",
- Cluster of Excellence EXC 1077/1 Hearing4all.

I'd like to thank Dr. Stephan D. Ewert and Prof. Dr. Steven van de Par for
continuous support on the algorithm development and evaluation.

I'd like to thank all the people who tested earlier versions of the Software,
contributed to it, reported on bugs, or made suggestions for improvements:
Thomas Biberger, Oliver Buttler, Stephan D. Ewert, Jan-Hendrik Flessner, Nico
Goessling, Julian Grosse, Andreas Haeussler, Stefan Klockgether, Steven van de
Par, Josef Poppitz, Joachim Thiemann. (Sorry if I forgot someone here.)

All files in the directory ./BASE/EXTERNAL and all subdirectories are written
by other authors. Please see the file headers or license.txt files further
informations on licenses. In some cases I did code modifications. For details,
see the comments in the code. The authors are:

Daniyal Amir        OVERWRITE_MERGE.M
Robert Baumgartner  HOR2GEO.M
Piotr Majdak        GEO2HORPOLAR.M
Ofek Shilon         RANDORTHMAT.M (modified Version for RAZR: RANDORTHMAT3.M)
Todd Welti          SMOOTHNEW3.M

Furthermore, some supplementary functions for HRTF databases, located in
respective subfolders of ./BASE/HRTF, are written by the following authors:

Hendrik Kayser (kayser-database), Gunnar Geissler (mk2-database).

The unreverberated sound samples in ./BASE/EXTERNAL/SAMPLES are
part of the Oldenburg Sentence Test for AFC Software Package by Stephan Ewert,
Daniel Berg, Hoertech gGmbH. See also
BASE/EXTERNAL/SAMPLES/OLSA_SENTENCES_README.TXT.


changelog

--------------------------------------------------------------------------------
Note: "[!]" Indicates a change that
- changes the default behavior of RAZR or
- might introduce incompatibilities with earlier versions.
--------------------------------------------------------------------------------


Revision 0.93 (Dec. 2017)
(Note also the release notes for version 0.92 since it was no public release.)

* Coupled rooms supported now. For details, see RAZR, EXAMPLE_COUPLED_ROOMS,
  GET_COUPLED_ROOMS.
* Improved documentation in several files.
* [!] Syntax for wall material specification. Now a colon ":" is used as
  separator, i.e., "dbase:material". The period symbol is not used longer to
  prevent confusion with "struct.field" syntax. For details, see help razr.

Known issues:
* Absorption and reflection filters that are created internally might be
  instable if room.freq contains 125 Hz. Set the options op.plot_absFilters = 1
  and op.plot_reflFilters = 1 to check the created filters. In doubt, choose
  frequencies >= 250 Hz.


Revision 0.92 (Dec. 2017, no public release)

* [!] Fixed: Image source jitter is attenuated if it would cause an image source
  being closer to the receiver than the direct sound source.
* [!] op.ism_ISposRandFactor was replaced by op.ism_jitter_factor. Jittering now
  depends on ISM order instead of img src distance to receiver.
  op.ism_randFactorsInCart was renamed to op.ism_jitter_type and takes string
  values now.
* apply_rir: New parameters 'samples' and 'win'. See help apply_rir for details.
* schroeder_rt: plots EDC automatically if nargout == 0.
* [!] New handling of wall materials. Own databases for absorption coefficients
  can now be created and used. For syntax see help razr.
* [!] New handling of configuration. Several cfg files can be stored in parallel
  now. The function select_razr_cfg picks the desired one. See doc of
  "razr_cfg_default" and "select_razr_cfg" for details.
* The SOFA format (sofaconventions.org) for HRTFs is supported now.
* [!] New syntax for adding unsupported HRTF databases (non-SOFA format). See
  README for details.
* New functions: sop, smat. Search tools for option names and absorbing
  materials, respectively.
* ISM needs less memory now.
* [!] RIR length is now specified in samples instead of seconds. Therefore, the
  options "op.tlen" and "op.tlen_max" were replaced by new options "op.len" and
  "op.len_max", respectively.
* [!] The automatically chosen RIR length is now set to the maximum of
  the estimated T60 over frequencies instead of the doubled maximum. However,
  this behavior can be controlled using the new option "op.len_rt_factor".


Revision 0.91 (Apr. 2017)

* New function plot_ispos to plot image source positions
* The razr output "ism_data" now contains more fields
* scene:
    - Plots room even if srcpos/recpos/recdir are not existing
    - Colors of sources and receivers can now be specified
    - For multiple sources/receivers, labels have now the same color
    - Structure of returned handles is more clear now
* New function plot_freqrsp: Replaces plotFrqRsp
* Some code cleanup and refactoring.
* Some minor fixes
* schroeder_rt:
    - can plot the reverberation time against frequencies now
    - new option to plot the linear EDC fit
* [!] No random jitter on 1st-order img-src positions. New option to explicitly
  control that threshold order: op.ism_rand_start_order.


Revision 0.90 (Nov. 2016)

* New options for scene: hide_labels, labeloffset
* Disabled tone correction filter on direct sound
* Introduced spherical head model. The according new spat_mode 'shm' is now the
  default instead of 'ild'
* razr addpath/rmpath ignores .git folder now
* razr_cfg is now only created for razr addpath (not rmpath)
* "razr" is printed in lower case letters for razr short help
* apply_hrtf: replaced feval by eval for calling scripts (warning in older
  Matlab versions)


Revision 0.11 (Oct. 2016)

* Simpler folder structure
* Changed way of including unsupported HRTF databases (see README.txt)
* Multiple sound sources and receivers are supported now (see
  example_multiple_src.m). Updated analysis tools: scene, plot_ir, plot_irspec.


razr.m
function varargout = razr(varargin)
% RAZR - Top-level function for RAZR engine
%
% Usage:
%   [ir, ism_setup, fdn_setup, ism_data] = RAZR(room)
%   [ir, ism_setup, fdn_setup, ism_data] = RAZR(room, op)
%   [ir, ism_setup, fdn_setup, ism_data] = RAZR(rooms, adj)
%   [ir, ism_setup, fdn_setup, ism_data] = RAZR(rooms, adj, op)
%   [ir, ism_setup, fdn_setup, ism_data] = RAZR(__, Name, Value)
% or:
%   RAZR            Displays short help
%   RAZR help       Displays complete help
%   RAZR about      Displays program infos
%   RAZR addpath    Adds the RAZR root directory and required folders to path
%   RAZR rmpath     Removes the the all added folders from search path
%   RAZR demo       Starts the RAZR demo (writes a wav file to RAZR's root dir.)
%
% Input:
%   room    Structure specifying a shoebox shaped room containing a sound source
%           and a receiver. Required fields are:
%
%           - boxsize       Vector [x, y, z], room dimensions in m
%           - srcpos        Vector [x, y, z], source position in m
%           - recpos        Vector [x, y, z], receiver position in m
%           - recdir        Vector [az, el], receiver orientation in degrees
%
%           Multiple source-receiver positions can be specified by using
%           matrices instead of row vectors. The number of srcpos, recpos and
%           recdir (i.e., number of rows) must either match, or only one of them
%           must be larger than 1.
%
%           - materials     Wall surface materials. Can be specified in
%                           different ways:
%                           - A cell array of 6 material strings. The order of
%                             wall surfaces is: {-z; -y; -x; +x; +y; +z}.
%                             Material strings must be in the format
%                             "dbase:mymaterial", where "dbase" is the name of
%                             an absorbing-material database that contains ab-
%                             sorption coefficients stored under "mymaterial".
%                             An example is given in GET_ROOM_L. Databases are
%                             stored in files GET_ABSCOEFF_<DBASE>.M. Own data-
%                             bases can be created in the same format and will
%                             automatically be recognized. An example database
%                             is: GET_ABSCOEFF_HALL.
%                           - A row vector containing absorption coefficients
%                             matching the frequencies stored in room.freq.
%                             Same values for all walls.
%                           - A column vector containing six frequency-indepen-
%                             dent absorption coefficients for each wall. Order
%                             of wall surfaces: [-z; -y; -x; +x; +y; +z].
%                           - A matrix containing frequency-dependent absorption 
%                             coefficients for six walls. Frequencies along row,
%                             walls along column. Order of walls: see above.
%                           - A scalar number specifying one absorption
%                             coefficient, same for all walls and frequencies.
%                           Internally, a new field "abscoeff" will be creaated
%                           being a matrix containing frequency dependent
%                           absorption coefficients for all room surfaces.
%           Instead of materials, a desired reverberation time can be specified:
%
%           - t60           Freq. dependent desired reverberation time in sec.
%
%           Note: If both materials and t60 are fields of a room, materials will
%           be used and t60 will be ignored.
%
%           - freq          Frequency base (in Hz) for materials (or t60) as
%                           octave band center frequencies. Optional, if
%                           materials are specified as cell array of strings.
%                           Default: [250, 500, 1e3, 2e3, 4e3]
%
%           - TCelsius      Room temperature in Celsius. It will influence the
%                           speed of sound. See also SPEEDOFSOUND. If not
%                           specified, TCelsius defaults to 20.
%
%           Some example rooms are defined in EXAMPLES/GET_ROOM_*.M.
%
%   rooms   Vector of rooms. To simulate two coupled rooms, each room must con-
%           tain an extra field: a door. The following fields are required:
%
%           - door.name     Name of the door as a string.
%           - door.wall     ID of the wall where the door is placed. Must be a
%                           value out of [-2, -1, 1, 2] corresponding to
%                           [-y, -x, +x, +y]
%           - door.pos      [x/y, z] position (lower vertex) at specified wall
%           - door.size     [width, height]
%
%           door can be a vector of multiple of these structures. However,
%           simulation is supported only for one door up to now.
%           The room vector must in total contain one source and one receiver
%           that are in the same room or in two neighboring rooms. For a room
%           that contains no src/rec the respective field must be an empty array.
%           For an example, see GET_COUPLED_ROOMS.M.
%           If length(rooms) > 2, those two rooms are taken into account, which
%           contain the source and the receiver. If source and receiver are
%           inside the same room and length(rooms) > 2, that adjacent room with
%           the highest estimated reverberation time will be taken into account.
%           Note: If a room vector is passed, adj must be passed, too (see
%           below).
%           An example to simulate two coupled rooms is provided in
%           EXAMPLE_COUPLED.M.
%
%   op      Structure specifying options for RIR synthesis (optional, if empty
%           or not specified, default options are applied. All avilable options
%           are listed in GET_DEFAULT_OPTIONS). Examples how to use options are
%           given in EXAMPLES/EXAMPLE_OPTIONS.M.
%
%   adj     Cell array containing adjacency specification for coupled rooms. For
%           specifications, names of rooms and doors are used.
%           Example:
%
%           adj = {...
%               'roomA_name', 'door_A1', 'roomB_name', 'door_B1'; ...
%               'roomA_name', 'door_A2', 'roomC_name', 'door_C1'}
%
%           This specifies a connection from room 'roomA_name' via door
%           'door_A1' to room 'roomB' via door 'door_B1' and a connection from
%           room 'roomA' via door 'door_A2' to room 'roomC' via door 'door_C1'.
%           An example to simulate two coupled rooms is provided in
%           EXAMPLE_COUPLED.M.
%
% Name-Value-pair arguments:
%   May be all fieldnames and values of options structure
%   (see GET_DEFAULT_OPTIONS).
%
% Output:
%   ir          Room impulse response as strcuture, containing (at least) the
%               following fields:
%
%               - sig           Time signal of impulse response
%               - fs            Sampling rate in Hz
%               - start_spl     Time sample at which the direct sound appears
%               - name          Name of impulse response (same as op.rirname)
%
%               Additional fields may be created, depnding on specified options.
%
%   ism_setup   Structure containing ISM setup created by GET_ISM_SETUP
%   fdn_setup   Structure containing FDN setup created by GET_FDN_SETUP
%   ism_data    Structure containing geometric data for image sources
%
% Example:
%   See EXAMPLE_DEFAULT, EXAMPLE_OPTIONS, EXAMPLE_MULTIPLE_SRC,
%   EXAMPLE_COUPLED_ROOMS
%
% Note:
%   For most usages, RAZR's root directory and all subfolders will be temporally
%   added to your MATLAB search path. You can remove them using >> razr rmpath
%
% Please see also the README and the changelog.
%
% See also: example_*, GET_ROOM_L, SCENE, SOP, GET_DEFAULT_OPTIONS

%------------------------------------------------------------------------------
% RAZR engine for Mathwork's MATLAB
%
% Version 0.93
%
% Author(s): Torben Wendt
%
% Copyright (c) 2014-2017, Torben Wendt, Steven van de Par, Stephan Ewert,
% University Oldenburg, Germany.
%
% This work is licensed under the
% Creative Commons Attribution-NonCommercial-NoDerivs 4.0 International
% License (CC BY-NC-ND 4.0).
% To view a copy of this license, visit
% http://creativecommons.org/licenses/by-nc-nd/4.0/ or send a letter to
% Creative Commons, 444 Castro Street, Suite 900, Mountain View, California,
% 94041, USA.
%------------------------------------------------------------------------------


if nargin == 0
    help_str = help('razr.m');
    idx_start = strfind(help_str, 'Usage:');
    idx_end = strfind(help_str, 'Input:');
    help_str = help_str((idx_start - 2):(idx_end - 1));
    help_str = strrep(help_str, 'RAZR', 'razr');
    disp(help_str);
    
elseif isstruct(varargin{1})
    % razr(room, …)
    
    razr_addpath(true);
    create_user_cfg;
    adj_exists = false;
    room = varargin{1};
    varargin(1) = [];
    
    if ~isempty(varargin)
        if iscell(varargin{1})
            % remaining varargins: adj, …
            adj = varargin{1};
            adj_exists = true;
            varargin(1) = [];
        end
        
        if ~isempty(varargin)
            if isstruct(varargin{1})
                % remaining varargins: op, …
                op = varargin{1};
                varargin(1) = [];
            else
                op = struct;
            end
            
            if ~isempty(varargin)
                % remaining varargins: Name, Value, …
                if mod(length(varargin), 2)
                    error('Wrong pattern of input parameters. See help razr.');
                end
                
                % pack cell arrays into cells, such that they are transformed
                % correctly to a struct:
                idx_cell = find(cellfun(@iscell, varargin));
                for idx = idx_cell
                    varargin(idx) = {varargin(idx)};
                end
                
                op = overwrite_merge(struct(varargin{:}), op, 1, 1);
            end
        else
            % razr(room, adj)
            op = struct;
        end
    else
        % razr(room)
        op = struct;
    end
    if length(room) > 1 && ~adj_exists
        error('Adjacency specification not passed to razr.');
    end
    
    if adj_exists
        [varargout{1}, varargout{4}, varargout{2}, varargout{3}] = ...
            create_crir(room, adj, op);
    else
        [varargout{1}, varargout{2}, varargout{3}, varargout{4}] = ...
            create_rir(room, op);
    end
    
else
    switch varargin{1}
        case 'help'
            help razr.m
        case 'addpath'
            razr_addpath(true);
            create_user_cfg;
        case 'rmpath'
            razr_addpath(false);
        case 'about'
            % prints first section of README:
            fname = fullfile(get_razr_path, 'README');
            % assume that email address is the last line to be printed:
            pattern = 'oldenburg.de';
            readmetext = fileread(fname);
            idx_pattern = strfind(readmetext, pattern);
            disp(readmetext(1:(idx_pattern + length(pattern))));
        case 'demo'
            razr_addpath(true);
            create_user_cfg;
            razr_demo;
        otherwise
            error('Argument unknown: %s. Type "razr help" to see all options.', ...
                varargin{1});
    end
end
